src/datastore/postgres-store.ts

```diff
@@ -183,11 +194,48 @@ export class PgDataStore extends (EventEmitter as { new (): DataStoreEventEmitte
     this.pool = pool;
   }
 
+  async getChainTipHeight(
+    client: ClientBase
+  ): Promise<{ blockHeight: number; blockHash: string; indexBlockHash: string }> {
+    const currentTipBlock = await client.query<{
+      block_height: number;
+      block_hash: Buffer;
+      index_block_hash: Buffer;
+    }>(
+      `
+      SELECT block_height, block_hash, index_block_hash
+      FROM blocks
+      WHERE canonical = true AND block_height = (SELECT MAX(block_height) FROM blocks)
+      `
+    );
+    const height = currentTipBlock.rows[0]?.block_height ?? 0;
+    return {
+      blockHeight: height,
+      blockHash: bufferToHexPrefixString(currentTipBlock.rows[0]?.block_hash ?? Buffer.from([])),
+      indexBlockHash: bufferToHexPrefixString(
+        currentTipBlock.rows[0]?.index_block_hash ?? Buffer.from([])
+      ),
+    };
+  }
+
   async update(data: DataStoreUpdateData): Promise<void> {
     const client = await this.pool.connect();
     try {
       await client.query('BEGIN');
-      await this.handleReorg(client, data.block);
+      const chainTip = await this.getChainTipHeight(client);
+      await this.handleReorg(client, data.block, chainTip.blockHeight);
+      // If the incoming block is not of greater height than current chain tip, then store data as non-canonical.
+      if (data.block.block_height <= chainTip.blockHeight) {
+        data.block.canonical = false;
+        data.txs.forEach(tx => {
+          tx.tx.canonical = false;
+          tx.stxEvents.forEach(e => (e.canonical = false));
+          tx.ftEvents.forEach(e => (e.canonical = false));
+          tx.nftEvents.forEach(e => (e.canonical = false));
```

Comment:
isn't it safer to use `.map` and do it immutably everywhere?

Reply:
Good call, done